File: apps/ios/Packages/TrackEvolutionKit/Sources/TrackEvolutionKit/Models/Event.swift
import Foundation

/// One item of an event's prep checklist.
public struct ChecklistItem: Codable, Hashable, Sendable {
    public var text: String
    public var done: Bool

    public init(text: String, done: Bool) {
        self.text = text
        self.done = done
    }
}

/// Track conditions for an event. Mirrors `CONDITIONS` in `src/lib/validate.ts`.
///
/// Deliberately a `RawRepresentable` struct rather than a Swift `enum`: a value
/// the server adds later must not fail the decode of a whole event on an app
/// that predates it. Use `Conditions.all` to drive a picker.
public struct Conditions: RawRepresentable, Codable, Hashable, Sendable {
    public let rawValue: String
    public init(rawValue: String) { self.rawValue = rawValue }

    public static let dry = Conditions(rawValue: "dry")
    public static let damp = Conditions(rawValue: "damp")
    public static let wet = Conditions(rawValue: "wet")
    public static let mixed = Conditions(rawValue: "mixed")

    public static let all: [Conditions] = [.dry, .damp, .wet, .mixed]
}

/// A track day. The canonical shape is `ComputedEvent` in `src/lib/stats.ts`
/// (`EVENT_SELECT` in `src/db.ts` for the columns).
///
/// Optionality here is load-bearing and mirrors `withComputed`:
/// - `bestMs` is nil when the event has neither a manual best nor any laps.
/// - `consistency` is nil below 3 laps — **not** zero.
/// - `hours` is never nil; the server already rounded it to 1dp.
/// - `checklist` is nil when absent *or* malformed (`parseChecklist` degrades
///   rather than throwing).
/// `lap_avg`/`lap_avg_sq` are stripped by `withComputed` and so are absent here.
public struct Event: Codable, Hashable, Sendable, Identifiable {
    public var id: Int
    public var trackId: Int
    public var trackName: String
    /// ISO `yyyy-mm-dd`, the format every date column stores.
    public var startDate: String
    public var days: Int
    public var club: String?
    public var runGroup: String?
    public var car: String?
    public var vehicleId: Int?
    public var notes: String?
    public var conditions: Conditions?
    /// Ambient temperature in °F — a whole number (`isValidTemp`).
    public var tempF: Int?
    public var checklist: [ChecklistItem]?
    /// Manually entered best lap, independent of logged laps.
    public var bestTimeMs: Int?
    /// Manual override of the on-track hours estimate.
    public var trackHours: Double?
    public var updatedAt: Int
    public var lapBestMs: Int?
    public var lapCount: Int
    public var sessionCount: Int
    /// `min(bestTimeMs, lapBestMs)` — the number the UI shows as *the* best.
    public var bestMs: Int?
    /// Coefficient of variation of lap times; nil below 3 laps.
    public var consistency: Double?
    /// On-track hours: the override, else `max(days × 2h, logged lap time)`.
    public var hours: Double

    public enum CodingKeys: String, CodingKey {
        case id
        case trackId = "track_id"
        case trackName = "track_name"
        case startDate = "start_date"
        case days
        case club
        case runGroup = "run_group"
        case car
        case vehicleId = "vehicle_id"
        case notes
        case conditions
        case tempF = "temp_f"
        case checklist
        case bestTimeMs = "best_time_ms"
        case trackHours = "track_hours"
        case updatedAt = "updated_at"
        case lapBestMs = "lap_best_ms"
        case lapCount = "lap_count"
        case sessionCount = "session_count"
        case bestMs = "best_ms"
        case consistency
        case hours
    }
}

/// `GET /api/events/:id` — an event plus its sessions and per-day setup sheets.
///
/// The event fields are decoded by `Event` from the same container, so the two
/// can never drift apart.
public struct EventDetail: Codable, Hashable, Sendable, Identifiable {
    public var event: Event
    public var sessions: [Session]
    public var setups: [Setup]

    public var id: Int { event.id }

    public enum CodingKeys: String, CodingKey {
        case sessions, setups
    }

    public init(from decoder: any Decoder) throws {
        event = try Event(from: decoder)
        let c = try decoder.container(keyedBy: CodingKeys.self)
        sessions = try c.decode([Session].self, forKey: .sessions)
        setups = try c.decode([Setup].self, forKey: .setups)
    }

    public func encode(to encoder: any Encoder) throws {
        try event.encode(to: encoder)
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(sessions, forKey: .sessions)
        try c.encode(setups, forKey: .setups)
    }
}

File: apps/android/core/src/main/kotlin/app/trackevolution/core/model/Event.kt
package app.trackevolution.core.model

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonEncoder
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject

/** One item of an event's prep checklist. */
@Serializable
public data class ChecklistItem(
    val text: String,
    val done: Boolean,
)

/**
 * Track conditions for an event. Mirrors `CONDITIONS` in `src/lib/validate.ts`.
 *
 * Deliberately a value class over `String` rather than a Kotlin `enum`: a value
 * the server adds later must not fail the decode of a whole event on an app that
 * predates it. Use [Conditions.all] to drive a picker.
 */
@Serializable
@JvmInline
public value class Conditions(public val rawValue: String) {
    public companion object {
        public val DRY: Conditions = Conditions("dry")
        public val DAMP: Conditions = Conditions("damp")
        public val WET: Conditions = Conditions("wet")
        public val MIXED: Conditions = Conditions("mixed")

        public val all: List<Conditions> = listOf(DRY, DAMP, WET, MIXED)
    }
}

/**
 * A track day. The canonical shape is `ComputedEvent` in `src/lib/stats.ts`
 * (`EVENT_SELECT` in `src/db.ts` for the columns).
 *
 * Optionality here is load-bearing and mirrors `withComputed`:
 *  - [bestMs] is null when the event has neither a manual best nor any laps.
 *  - [consistency] is null below 3 laps — **not** zero.
 *  - [hours] is never null; the server already rounded it to 1dp.
 *  - [checklist] is null when absent *or* malformed (`parseChecklist` degrades
 *    rather than throwing).
 *
 * `lap_avg`/`lap_avg_sq` are stripped by `withComputed` and so are absent here.
 */
@Serializable
public data class Event(
    val id: Int,
    @SerialName("track_id") val trackId: Int,
    @SerialName("track_name") val trackName: String,
    /** ISO `yyyy-mm-dd`, the format every date column stores. */
    @SerialName("start_date") val startDate: String,
    /**
     * Track days, **fractional**: the column is `days REAL` and the web form's
     * input steps by 0.5, so a half-day event is 0.5 and a Saturday-plus-Sunday-
     * morning weekend is 1.5. Modelling this as `Int` fails the decode of the
     * entire events list for anyone who has ever logged one.
     */
    val days: Double,
    val club: String? = null,
    @SerialName("run_group") val runGroup: String? = null,
    val car: String? = null,
    @SerialName("vehicle_id") val vehicleId: Int? = null,
    val notes: String? = null,
    val conditions: Conditions? = null,
    /** Ambient temperature in °F — a whole number (`isValidTemp`). */
    @SerialName("temp_f") val tempF: Int? = null,
    val checklist: List<ChecklistItem>? = null,
    /** Manually entered best lap, independent of logged laps. */
    @SerialName("best_time_ms") val bestTimeMs: Int? = null,
    /** Manual override of the on-track hours estimate. */
    @SerialName("track_hours") val trackHours: Double? = null,
    /**
     * Epoch milliseconds, trigger-maintained (migration 0011) and the input to
     * the offline cache's staleness check. `Long`, not `Int`: Kotlin's `Int` is
     * 32 bits and an epoch-ms timestamp passed that in 1970. (Swift's `Int` is
     * 64-bit on every platform we ship, which is why NS-04 can spell it `Int`
     * and still mean the same thing.)
     */
    @SerialName("updated_at") val updatedAt: Long,
    @SerialName("lap_best_ms") val lapBestMs: Int? = null,
    @SerialName("lap_count") val lapCount: Int,
    @SerialName("session_count") val sessionCount: Int,
    /** `min(bestTimeMs, lapBestMs)` — the number the UI shows as *the* best. */
    @SerialName("best_ms") val bestMs: Int? = null,
    /** Coefficient of variation of lap times; null below 3 laps. */
    val consistency: Double? = null,
    /** On-track hours: the override, else `max(days × 2h, logged lap time)`. */
    val hours: Double,
)

/**
 * `GET /api/events/:id` — an event plus its sessions and per-day setup sheets.
 *
 * The server returns one flat object: the event's own columns sit alongside
 * `sessions` and `setups`. [EventDetailSerializer] splits them back apart so the
 * event fields are decoded by [Event] itself and the two can never drift.
 */
@Serializable(with = EventDetailSerializer::class)
public data class EventDetail(
    val event: Event,
    val sessions: List<Session>,
    val setups: List<Setup>,
)

/**
 * Flattens/unflattens [EventDetail] against the server's single object.
 *
 * JSON-only by construction: it reaches for [JsonDecoder]/[JsonEncoder] because
 * splitting a keyed object into "these keys are an Event, those two are not"
 * has no meaning in a non-JSON format, and this type only ever crosses the wire
 * as JSON.
 */
public object EventDetailSerializer : KSerializer<EventDetail> {
    private val sessionsSerializer = ListSerializer(Session.serializer())
    private val setupsSerializer = ListSerializer(Setup.serializer())

    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("EventDetail") {
        element("event", Event.serializer().descriptor)
        element("sessions", sessionsSerializer.descriptor)
        element("setups", setupsSerializer.descriptor)
    }

    override fun deserialize(decoder: Decoder): EventDetail {
        val input = decoder as? JsonDecoder
            ?: throw IllegalStateException("EventDetail can only be decoded from JSON")
        val root = input.decodeJsonElement().jsonObject
        val eventFields = JsonObject(root.filterKeys { it != "sessions" && it != "setups" })
        return EventDetail(
            event = input.json.decodeFromJsonElement(Event.serializer(), eventFields),
            sessions = input.json.decodeFromJsonElement(
                sessionsSerializer,
                root["sessions"] ?: error("event detail has no sessions"),
            ),
            setups = input.json.decodeFromJsonElement(
                setupsSerializer,
                root["setups"] ?: error("event detail has no setups"),
            ),
        )
    }

    override fun serialize(encoder: Encoder, value: EventDetail) {
        val output = encoder as? JsonEncoder
            ?: throw IllegalStateException("EventDetail can only be encoded as JSON")
        val fields = LinkedHashMap<String, kotlinx.serialization.json.JsonElement>(
            output.json.encodeToJsonElement(Event.serializer(), value.event).jsonObject,
        )
        fields["sessions"] = output.json.encodeToJsonElement(sessionsSerializer, value.sessions)
        fields["setups"] = output.json.encodeToJsonElement(setupsSerializer, value.setups)
        output.encodeJsonElement(JsonObject(fields))
    }
}
